pkg/registry/discovery/endpointslice/strategy.go

```diff
@@ -186,11 +189,40 @@ func dropTopologyOnV1(ctx context.Context, oldEPS, newEPS *discovery.EndpointSli
 			return
 		}
 
+		oldDeprecatedTopologyNodeNames := getDeprecatedTopologyNodeNames(oldEPS)
+
 		for i := range newEPS.Endpoints {
 			ep := &newEPS.Endpoints[i]
 
-			//Silently clear out DeprecatedTopology
-			ep.DeprecatedTopology = nil
+			if ep.DeprecatedTopology != nil {
+				newTopologyNodeName, ok := ep.DeprecatedTopology[v1.LabelHostname]
+				if ep.NodeName == nil && oldDeprecatedTopologyNodeNames != nil && ok && oldDeprecatedTopologyNodeNames.Has(newTopologyNodeName) {
```

Comment:
no need to check `oldDeprecatedTopologyNodeNames != nil`, Has on nil string set returns false